### PR TITLE
Add additional context to configuration mutation deprecations

### LINKED ISSUE #31505: Find when a configuration is resolved
### Expected Behavior

I see deprecation messages like the following in my build scan:

```
Mutating the dependencies of configuration '<project>' after it has been resolved or consumed. This behavior has been deprecated.
    This will fail with an error in Gradle 9.0.
    After a Configuration has been resolved, consumed as a variant, or used for generating published metadata, it should not be modified.
```

This comes with a corresponding stack trace. However, I would also like to see where in my build the configuration has been resolved so that I can decide which location needs to be changed. Is this possible?

### Current Behavior (optional)

_No response_

### Context

I am attempting to prepare some builds for Gradle 9.

### PR DESCRIPTION
**Pull request by jvandort**
We added deprecations when a configuration that has been observed through resolution, consumption, or publishing -- of itself or a child configuration -- would become an error in Gradle 9.0

The deprecation is triggered when performing the mutation, but is 'enabled' when the configuration is observed. Since these happen at different times, the stack trace from the deprecation does not provide any details on how to avoid the prior observation.

This commit attempts to capture some context during observation that can be exposed in the deprecation warning.

We can now tell, upon mutation, that a configuration was observed via publishing, resolution, or consumption as a variant.

One downside to this approach, is we cannot tell _which_ resolution consumed a given configuration as a variant. This would be quite difficult to solve, or at least would require a lot of wiring that is not worth it.

For users who cannot resolve their deprecationw with this additional context, we should guide them to debugging Gradle to determine exactly how their configuration became observed

Fixes #31505
### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test apt

**Comment by ov7a** [comment]
🎉 I think it's great that we can provide some context to help users fix the deprecation warning.

Regarding the message:
```
Mutating the dependencies of configuration ':conf' after it has been resolved or consumed. This behavior has been deprecated. This will fail with an error in Gradle 9.0. Resolved configuration ':conf'. After a Configuration has been resolved, consumed as a variant, or used for generating published metadata, it should not be modified. Consult the upgrading guide for further information: https://docs.gradle.org/current/userguide/upgrading_version_8.html#mutate_configuration_after_locking")
```

I'd suggest to rephrase the context from
```
Resolved configuration ':conf'.
```
to something more verbose like this:
```
This issue was triggered because configuration `:conf` was resolved beforehand.
```
And in child case:
```
This issue was triggered because a child configuration was observed: configuration ':compileClasspath' was resolved beforehand.
```

**Comment by jvandort** [comment]
I've made some changes to improve the flow of the deprecation message. WDYT?

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/api/ConfigurationDefaultsIntegrationTest.groovy [hunk 1]
```diff
@@ -241,39 +241,6 @@ include 'consumer', 'producer'
         }
     }
 
-    def "can use beforeResolve to specify default dependencies"() {
```
Comment (jvandort) [inline platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/api/ConfigurationDefaultsIntegrationTest.groovy:244]: A very similar test was already present in `BeforeResolveIntegrationTest`

**Comment by jvandort** [comment]
@bot-gradle test this

**Review by ov7a — APPROVED** [review]

**Pull request merged**